zilliqa/src/exec.rs

```diff
@@ -673,23 +676,101 @@ impl State {
         value: u128,
     ) -> Result<u64> {
         let gas_price = gas_price.unwrap_or(GAS_PRICE);
-        let gas = gas.unwrap_or(BLOCK_GAS_LIMIT);
 
+        let mut right = gas.unwrap_or(MAX_EVM_GAS_LIMIT).0;
+        let mut left = lower_bound_gas_estimate(to_addr, &data);
```

Comment:
What if we instead generating the lower bound by first running the transaction with `gas_limit = BLOCK_GAS_LIMIT` and took the `gas_used` from the result?

Reply:
In fact, I think you can just set `left`/`min` to `gas_used` from your "Check if estimation succeeds with the highest possible gas" call below?